File: movegen/knight_castling.c
// knight_castling.c
// Genenerates knight positions that are impediments to castling

#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include "knightmove.h"

int main(int argc, char **argv){

	fprintf(stdout,"// knight_castling.h\n// constants for where knights may attack castling squares\n// In the following, Knight_Castling[0] = WKING, Knight_Castling[1] = WQUEEN, Knight_Castling[2] = BKING, ...[3] = BQUEEN\n\n#define KC_WKING 0\n#define KC_WQUEEN 1\n#define KC_BKING 2\n#define KC_BQUEEN 3\n\n");
	uint64_t Knight_Castling[4] = {
	(knightmoves[4 + 8*0] | knightmoves[5 + 8*0] | knightmoves[6 + 8*0]),
	(knightmoves[2 + 8*0] | knightmoves[3 + 8*0] | knightmoves[4 + 8*0]),
	(knightmoves[4 + 8*7] | knightmoves[5 + 8*7] | knightmoves[6 + 8*7]),
	(knightmoves[2 + 8*7] | knightmoves[3 + 8*7] | knightmoves[4 + 8*7]),
	};


	fprintf(stdout, "const uint64_t Knight_Castling[4] = {\n");
	for(int i = 0; i < 4; i++){
		fprintf(stdout, "0x%llx,\n",(long long unsigned) Knight_Castling[i]);
	}
	fprintf(stdout,"};\n");
        return 0;
}

File: src/movegen/knight_castling.py
#!/usr/bin/python
# generates knight moves that attack castling squares

print "// knight_castling.h"
print "// constants for where knights may attack castling squares"
print "// In the following, Knight_Castling[0] = WKING, Knight_Castling[1] = WQUEEN, Knight_Castling[2] = BKING, ...[3] = BQUEEN"
print ""
print "#define KC_WKING 0"
print "#define KC_WQUEEN 1"
print "#define KC_BKING 2"
print "#define KC_BQUEEN 3"
print ""
print "const uint64_t Knight_Castling[4] = {"
# the code below is wrong because it double counts, and += 2**(.) is not the same as bitwise and
# for i in [0, 7]:
# 	result = 0
# 	for j in [4, 5, 6]: # KINGSIDE
# 		n = i * 8 + j
# 		if (i + 1 <= 7) and (j + 2 <= 7):
# 			result += 2**(n + 10)
# 		if (i + 1 <= 7) and (j - 2 >= 0):
# 			result += 2**(n + 6)
# 		if (i - 1 >= 0) and (j + 2 <= 7):
# 			result += 2**(n - 6)
# 		if (i - 1 >= 0) and (j - 2 >= 0):
# 			result += 2**(n - 10)
# 
# 		if (i + 2 <= 7) and (j + 1 <= 7):
# 			result += 2**(n + 17)
# 		if (i + 2 <= 7) and (j - 1 >= 0):
# 			result += 2**(n + 15)
# 		if (i - 2 >= 0) and (j + 1 <= 7):
# 			result += 2**(n - 15)
# 		if (i - 2 >= 0) and (j - 1 >= 0):
# 			result += 2**(n - 17)
# 	print str(hex(result)) + ","
# 
# 	result = 0
# 	for j in [2, 3, 4]: #QUEENSIDE
# 		n = i * 8 + j
# 		if (i + 1 <= 7) and (j + 2 <= 7):
# 			result += 2**(n + 10)
# 		if (i + 1 <= 7) and (j - 2 >= 0):
# 			result += 2**(n + 6)
# 		if (i - 1 >= 0) and (j + 2 <= 7):
# 			result += 2**(n - 6)
# 		if (i - 1 >= 0) and (j - 2 >= 0):
# 			result += 2**(n - 10)
# 
# 		if (i + 2 <= 7) and (j + 1 <= 7):
# 			result += 2**(n + 17)
# 		if (i + 2 <= 7) and (j - 1 >= 0):
# 			result += 2**(n + 15)
# 		if (i - 2 >= 0) and (j + 1 <= 7):
# 			result += 2**(n - 15)
# 		if (i - 2 >= 0) and (j - 1 >= 0):
# 			result += 2**(n - 17)
# 	print str(hex(result)) + ","
# 
# print "};"
